### ISSUE TITLE
react/jsx-key should violate when key is potentially undefined

### ISSUE THREAD

**Issue by jroru** [jsx-eslint/eslint-plugin-react#3920]
```ts
interface Dog {
  id?: string;
  name: string;
}

const dogs: Dog[] = [
  { id: "spikey", name: "spike" },
  { name: "spot" },
] as const;

export const Dogs = () => (
  <div>
    {dogs.map((dog) => (
      <div key={dog.id}>{dog.name}</div>
    ))}
  </div>
);
````

The above code does not currently violate [jsx-key](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-key.md). With [type information](https://github.com/jsx-eslint/eslint-plugin-react/issues/3185) it could, and I believe it really should. Something that can potentially be undefined is unsuitable as a key.

**Comment by ljharb** [comment]
That's only true if more than one of the items has an undefined `id`.
